server/src/main/java/org/elasticsearch/index/IndexService.java

```diff
@@ -546,7 +555,8 @@ public synchronized IndexShard createShard(
                 circuitBreakerService,
                 snapshotCommitSupplier,
                 System::nanoTime,
-                indexCommitListener
+                indexCommitListener,
+                fieldHasValue
```

Comment:
Maybe that makes the comments verbose, but it'd be good to be very explicit that this set is shared across shards for a certain index that are allocated on the same node. This would need to support concurrency even within a single shard, but I wonder if there will be contention while indexing due to the fact that the data structure is shared across shards.